### ISSUE TITLE
How to inject SCTE35 splice private section

### ISSUE THREAD

**Issue by aikihook** [tsduck/tsduck#121]
Hi,
I'm trying to use tsp to produce a TS stream with PAT/PMT and single scte35 component.

**NOTE: There is no video so the PCR reference is 0x1FFF.**

I need to create a splice_null packet at startup and then later switch to splice_private command.
So I set up tsp to produce null bitrate stream @ 1Mbps, inject PAT, inject PMT
1) **tsp -b 1000000 -I null -P regulate -b 1000000**
2) **-P inject --poll-files pat.xml=500 -p 0 -i 332**
3) **-P inject --poll-files pmt.xml=500 -p 4660 -I 332**

-P { do something to splice in the scte35 section}
**- O scte35_stream.ts**

So far so good!

In my code I create a splice_section.bin file, that can holds two different types of splice e command:
 a) splice_null command
 b) splice_private command payload

BUT when I try to inject the splice.BIN section I hit a few problems. I tried the following:

4) FAILURE  using '**-P splice_inject splice_*.bin**'. Complains that there is no PCR or PTS. I tried all options of specifying i.e. '--pts-pid 8191' to no avail.

5) FAILUE using '**-P inject --poll-files splice_section.bin=1000  -i 664**'. Complains "truncated section got 188 bytes expected 294. I checked and checked my splice_null packet its is perfectly created tableID=0xFC, section length exactly 17 bytes including CRC.

6) FAILURE/Partial success using '**-P mux -i 332 splice_section.bin**'. This works on MacOSX and I can rewrite over the the splice_section.bin with splice_private command data. The changed section gets switched into the output TS file stream correctly.

BUT on Windows 10 it seems '**-P mux splice_section.bin**' file is opened in an exclusive file mode and thus my running process code can not access the file to change the content.

I tried an alternative in my code to open the file in shared access and keep the file handle open - but then the 'mux' plugin appears to get blocked and output does not pass to scte35_stream.ts.

???

**Comment by lelegard** [comment]
Hi @aikihook 

Let's address the issues one by one

> I need to create a splice_null packet at startup and then later switch to splice_private command.

In practice, an SCTE 35 _section_ probably fits into one TS _packet_. But pay attention that we are talking about sections here, not packets. These are two distinct levels of transport. Mixing them may explain other issues here.

> FAILURE using '-P splice_inject splice_*.bin'. Complains that there is no PCR or PTS. I tried all options of specifying i.e. '--pts-pid 8191' to no avail.

The plugin `splice_inject` is used to handle SCTE 35 sections specifically. Except the null command, SCTE 35 is all about PTS synchronization. So, without PTS (or PCR as a fallback), the plugin does not know where to insert sections. To insert null commands in a stream of null packets, the generic table injection plugin `inject` is probably more appropriate.

>using '-P inject --poll-files splice_section.bin=1000 -i 664'. Complains "truncated section got 188 bytes expected 294. I checked and checked my splice_null packet its is perfectly created tableID=0xFC, section length exactly 17 bytes including CRC.

What does your .bin file contain? TS packets or sections? Given the message, I would think one TS packet. But the plugin `inject` needs either binary _section_ files or XML files. See the chapter on data formats in the TSDuck user's guide.

>using '-P mux -i 332 splice_section.bin'. This works on MacOSX and I can rewrite over the the splice_section.bin with splice_private command data. The changed section gets switched into the output TS file stream correctly.

Confirming my previous remark. The plugin `inject` inserts _sections_ while the plugin `mux` inserts _packets_.

> BUT on Windows 10 it seems '-P mux splice_section.bin' file is opened in an exclusive file mode and thus my running process code can not access the file to change the content.

Implicit file locking is the difference between Unix (including macOS) and Windows. 

For what you want to do, inserting a section that you create dynamically at any time, I suggest to use the plugin `inject` with option `--poll-files` and you create a _section_ file, not a TS file.

> I tried an alternative in my code to open the file in shared access and keep the file handle open - but then the 'mux' plugin appears to get blocked and output does not pass to scte35_stream.ts.

If you develop an application (and if C++ code is OK for you), I also suggest to look at the TSDuck library [here](https://tsduck.io/doxy/index.html) instead of using a `tsp` command. With this library, you can create, modify, analyse sections, tables, TS files, etc.

**Comment by aikihook** [comment]
Hi Thierry,
Thank you for your response.

First I must say that Tsduck is a most excellent suite of tools and saving me a lot of work re-inventing the world of MPEG2TS.

In my initial analysis and reading your very good documentation. I did think `inject` is the best option to use and is where I started trying to work with first off.  Yes I am using a 188 TS packet for the splice_null command - now I see why I get 'truncated section error'.

**Just a quick question, does XML processing `inject` module handle splice section data - i.e. like `splice_inject` module does ?**

However, I think I will work with binary sections that I form myself for now. The splice_private command  my application will receive is an encoded section with secret data.

So I will revert my code to simply forming the` binary section` - tableID 0xFC files and `inject` with `--poll-files`.

Just some extra information. My requirements dictate, I am constrained to using C#. In my code application I created a TsduckController console app and simply, execute `tsp` process passing the relevant `plugin` command line arguments. Thus, I am hoping I will not need to modify or extend the C++ library unless absolutely necessary.
